### ISSUE TITLE
[Epic] Dashboards Metadata Storage Decoupling

### ISSUE THREAD

**Issue by boktorbb-amzn** [opensearch-project/OpenSearch-Dashboards#1441]
The encompassing issue for all things Metadata Storage and decoupling it from OpenSearch

**Comment by ananzh** [comment]
Could you specify a bit 1) what are the Metadata 2) what they are used for 3) why we need to decouple? Thanks

**Comment by boktorbb-amzn** [comment]
> Could you specify a bit 1) what are the Metadata 2) what they are used for 3) why we need to decouple? Thanks

Sure. I've updated the description of the issue for more context.

**Comment by seraphjiang** [comment]
Add one few requests,  be able to allow admin who setup OSD to pick up whichever storage they want. 

Save meta to opensearch is still a option

Provide tool/feature to migrate meta data from one storage to another.

A few thinking but nice to have, not only decouple the storage but also make it a separate meta service outside of dahsboards implementation

**Comment by boktorbb-amzn** [comment]
> Add one few requests, be able to allow admin who setup OSD to pick up whichever storage they want.
> 
> Save meta to opensearch is still a option
> 
> Provide tool/feature to migrate meta data from one storage to another.
> 
> A few thinking but nice to have, not only decouple the storage but also make it a separate meta service outside of dahsboards implementation

Absolutely! Multiple storage options including OpenSearch is something we hope to work.

**Comment by kavilla** [comment]
[Groom]:

Please feel free to update any questions. And then figure out some scope and timelines.

**Comment by AMoo-Miki** [comment]
Metadata storage decoupling shouldn't be just about being able to use something other than OpenSearch. We need a simple replicating standalone key-value store to completely separate OSD from OS.

**Comment by hdhalter** [comment]
Does this need documentation?

**Comment by zengyan-amazon** [comment]
Closing — the core deliverables of this epic have been completed:

- **Design RFC:** PR #3954 (merged May 2023) — [Design for New Saved Object Service Interface for Custom Repository](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3954)
- **Implementation:** PR #4149 (merged Jun 2023) — [Adds Repository Factory Provider](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4149)

The Repository Factory Provider makes the `ISavedObjectRepository` implementation replaceable by plugin, decoupling Dashboards metadata storage from the `.opensearch_dashboards` index. Plugin authors can now provide custom repository implementations backed by any storage backend. Thank you!
